src/S3/Crypto/InstructionFileMetadataStrategy.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+namespace Aws\S3\Crypto;
+
+use \Aws\Crypto\MetadataStrategyInterface;
+use \Aws\Crypto\MetadataEnvelope;
+use \Aws\S3\S3Client;
+
+/**
+ * Stores and reads encryption MetadataEnvelope information in a file on Amazon
+ * S3.
+ *
+ * A file with the contents of a MetadataEnvelope will be created or read from
+ * alongside the base file on Amazon S3. The provided client will be used for
+ * reading or writing this object. A specified suffix (default of '.instruction'
+ * will be applied to each of the operations involved with the instruction file.
+ */
+class InstructionFileMetadataStrategy implements MetadataStrategyInterface
+{
+    const DEFAULT_FILE_SUFFIX = '.instruction';
+
+    private $client;
+    private $suffix;
+
+    /**
+     * @param S3Client $client Client for use in uploading the instruction file.
+     * @param string|null $suffix Optional override suffix for instruction file
+     *                            object keys.
+     */
+    public function __construct(S3Client $client, $suffix = null)
+    {
+        $this->suffix = empty($suffix)
+            ? self::DEFAULT_FILE_SUFFIX
+            : $suffix;
+        $this->client = $client;
+    }
+
+    /**
+     * Places the information in the MetadataEnvelope to a location on S3.
+     *
+     * @param MetadataEnvelope $envelope Encryption data to save according to
+     *                                   the strategy.
+     * @param array $args Starting arguments for PutObject, used for saving
+     *                    extra the instruction file.
+     *
+     * @return array Updated arguments for PutObject.
+     */
+    public function save(MetadataEnvelope $envelope, array $args)
+    {
+        $this->client->putObject([
```

Comment:
What if saving this instruction file fails? Is the ciphertext blob deleted from S3?

Reply:
The [`$strategy->save()` call](https://github.com/aws/aws-sdk-php/pull/1395/files#diff-a4b2e1f92c8fda8b31a82e07e90466e3R186) is made before the `PutObject` call for the ciphertext. The instruction file will remain on S3 if there is a failure after it has been uploaded, not automatically deleted. I'll add a note to the guide and update the doc block for this file.